tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -226,6 +226,288 @@ static bool HasCanonicalDimensionNumbers(
+  static constexpr StringRef kEllipsis = "...";
+
+  static bool CheckBatchHasEqualRank(size_t lhs_rank, StringRef lhs_loop,
+                                     size_t rhs_rank, StringRef rhs_loop,
+                                     size_t out_rank, StringRef out_loop);
+  static SmallVector<std::string> GetEinsumConfigAsVector(StringRef loop,
+                                                          size_t operand_rank);
+};
+
+// Definition of util const member variables.
+constexpr StringRef EinsumToLinalgConverter::kArrow;
+constexpr StringRef EinsumToLinalgConverter::kComma;
+constexpr StringRef EinsumToLinalgConverter::kEllipsis;
+
+// Convert the representation from string/vector<char> to vector<string>.
+// i.e ("abc") -> {"a", "b", "c"}. For cases with ellipsis with batch rank 3:
+// get loop_dim = f("ab...cde") = {"a","b","0","1","2","c","d","e"}
+SmallVector<std::string> EinsumToLinalgConverter::GetEinsumConfigAsVector(
+    StringRef loop, size_t operand_rank) {
+  SmallVector<std::string> loop_dim;
+  size_t pre_elip = loop.find(kEllipsis);
+  bool has_elip = pre_elip != std::string::npos;
+  if (!has_elip) pre_elip = loop.size();
+  // Add the dimension until the end or up to ellipsis if it exist.
+  for (int pre_elip_ind = 0; pre_elip_ind < pre_elip; pre_elip_ind++) {
+    loop_dim.push_back(loop.substr(pre_elip_ind, 1).str());
+  }
+  if (!has_elip) return loop_dim;
+  // Case where Ellipsis presence:
+  size_t non_batch_rank = loop.size() - kEllipsis.size();
+  size_t batch_rank = operand_rank - non_batch_rank;
+  // Add the batch dimension ("0",...,"N") where N is rank of batch into the
+  // loop.
+  for (int batch_ind = 0; batch_ind < batch_rank; batch_ind++) {
+    loop_dim.push_back(std::to_string(batch_ind));
+  }
+  // Add the dimension after ellipsis into the loop.
+  int post_elip = pre_elip + kEllipsis.size();
+  for (int post_elip_ind = post_elip; post_elip_ind < loop.size();
+       post_elip_ind++) {
+    loop_dim.push_back(loop.substr(post_elip_ind, 1).str());
+  }
+  return loop_dim;
+}
+
+// Returns true if all operand's batch has same rank.
+bool EinsumToLinalgConverter::CheckBatchHasEqualRank(
+    size_t lhs_rank, StringRef lhs_loop, size_t rhs_rank, StringRef rhs_loop,
+    size_t out_rank, StringRef out_loop) {
+  SmallVector<int, 3> batch_rank_vec;
+  if (lhs_rank != lhs_loop.size()) {
+    size_t lhs_batch_rank = lhs_rank - (lhs_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(lhs_batch_rank);
+  }
+  if (rhs_rank != rhs_loop.size()) {
+    size_t rhs_batch_rank = rhs_rank - (rhs_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(rhs_batch_rank);
+  }
+  if (out_rank != out_loop.size()) {
+    size_t out_batch_rank = out_rank - (out_loop.size() - kEllipsis.size());
+    batch_rank_vec.push_back(out_batch_rank);
+  }
+  bool batch_has_equal_rank = true;
+  if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                  batch_rank_vec.begin()) &&
```

Comment:
The PR is getting reverted because of a MSAN failure here. It may be that batch_rank_vec is empty so providing `batch_rank_vec.begin()+1` is UB.

Reply:
@joker-eph Is there a way I can recreate the MSAN test?

Reply:
@joker-eph I added a check in my fork [here](https://github.com/raikonenfnu/tensorflow/commit/4545360ef435b14bd2d487971d67b21b8ebc9635#diff-5f608b167d4dc35b869c40a5dafb3457312227aa75a1b1525fd9831ab831d58aR506). Should I open a new PR then?

Reply:
Yes you need a new pull-request

Reply:
@joker-eph sounds good! I made a new PR [here](https://github.com/tensorflow/tensorflow/pull/50811). Should I also change the name of the PR?